File: packages/honker-ex/test/phase_mantle_test.exs
defmodule PhaseMantleTest do
  use ExUnit.Case, async: false

  @candidates [
    "target/release/libhonker_ext.dylib",
    "target/release/libhonker_ext.so",
    "target/release/libhonker_extension.dylib",
    "target/release/libhonker_extension.so"
  ]

  @repo_root Path.expand("../../..", __DIR__)

  defp find_extension do
    Enum.find_value(@candidates, fn rel ->
      p = Path.join(@repo_root, rel)
      if File.exists?(p), do: p, else: nil
    end)
  end

  setup do
    case find_extension() do
      nil ->
        {:skip, "honker extension not built — run cargo build -p honker-extension --release"}

      ext ->
        dir = Path.join(System.tmp_dir!(), "honker-mantle-#{System.unique_integer([:positive])}")
        File.mkdir_p!(dir)
        db_path = Path.join(dir, "t.db")
        on_exit(fn -> File.rm_rf!(dir) end)
        {:ok, %{ext: ext, db_path: db_path}}
    end
  end

  test "schedule list round-trips fields", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    :ok =
      Honker.Scheduler.add(db,
        name: "recap",
        queue: "emails",
        schedule: "0 9 * * 1",
        payload: %{"team" => "premier-league"},
        priority: 3
      )

    :ok =
      Honker.Scheduler.add(db,
        name: "sync",
        queue: "syncs",
        schedule: "@every 1h",
        payload: nil
      )

    {:ok, rows} = Honker.Scheduler.list(db)
    assert length(rows) == 2
    recap = Enum.find(rows, &(&1["name"] == "recap"))
    assert recap["queue"] == "emails"
    assert recap["priority"] == 3
    assert recap["enabled"] == true
    assert Jason.decode!(recap["payload"])["team"] == "premier-league"
  end

  test "pause / resume idempotent", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    :ok =
      Honker.Scheduler.add(db,
        name: "a",
        queue: "q",
        schedule: "0 9 * * *",
        payload: nil
      )

    assert {:ok, true} = Honker.Scheduler.pause(db, "a")
    assert {:ok, false} = Honker.Scheduler.pause(db, "a")
    assert {:ok, false} = Honker.Scheduler.pause(db, "missing")

    {:ok, rows} = Honker.Scheduler.list(db)
    assert Enum.find(rows, &(&1["name"] == "a"))["enabled"] == false

    assert {:ok, true} = Honker.Scheduler.resume(db, "a")
    assert {:ok, false} = Honker.Scheduler.resume(db, "a")
    {:ok, rows} = Honker.Scheduler.list(db)
    assert Enum.find(rows, &(&1["name"] == "a"))["enabled"] == true
  end

  test "update mutates fields, no-op on empty, recomputes next_fire_at on cron change", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    :ok =
      Honker.Scheduler.add(db,
        name: "t",
        queue: "q",
        schedule: "0 9 * * *",
        payload: %{"v" => 1}
      )

    assert {:ok, true} = Honker.Scheduler.update(db, "t", payload: %{"v" => 99}, priority: 5)
    {:ok, rows} = Honker.Scheduler.list(db)
    row = Enum.find(rows, &(&1["name"] == "t"))
    assert Jason.decode!(row["payload"])["v"] == 99
    assert row["priority"] == 5

    before = row["next_fire_at"]
    assert {:ok, true} = Honker.Scheduler.update(db, "t", schedule: "*/5 * * * *")
    {:ok, rows} = Honker.Scheduler.list(db)
    row = Enum.find(rows, &(&1["name"] == "t"))
    assert row["cron_expr"] == "*/5 * * * *"
    refute row["next_fire_at"] == before

    assert {:ok, false} = Honker.Scheduler.update(db, "t", [])
    assert {:ok, false} = Honker.Scheduler.update(db, "missing", payload: %{})
  end

  test "queue cancel + get_job", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    {:ok, id} = Honker.Queue.enqueue(db, "emails", %{"to" => "alice@example.com"})

    {:ok, row} = Honker.Queue.get_job(db, id)
    assert row["queue"] == "emails"
    assert row["state"] == "pending"
    assert row["id"] == id

    assert {:ok, true} = Honker.Queue.cancel(db, id)
    assert {:ok, false} = Honker.Queue.cancel(db, id)
    assert {:ok, nil} = Honker.Queue.get_job(db, id)
    assert {:ok, nil} = Honker.Queue.claim_one(db, "emails", "worker-1")
  end

  test "cancel of processing invalidates ack", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    {:ok, id} = Honker.Queue.enqueue(db, "emails", %{"to" => "x"})
    {:ok, job} = Honker.Queue.claim_one(db, "emails", "worker-1")
    assert job.id == id

    assert {:ok, true} = Honker.Queue.cancel(db, id)
    assert {:ok, false} = Honker.Job.ack(db, job)
  end

  test "paused schedule does not emit on tick", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    :ok =
      Honker.Scheduler.add(db,
        name: "due",
        queue: "emails",
        schedule: "@every 1s",
        payload: %{"x" => 1}
      )

    Process.sleep(1100)
    assert {:ok, true} = Honker.Scheduler.pause(db, "due")

    {:ok, fires} = Honker.Scheduler.tick(db)
    assert fires == [], "paused schedule must not emit; got #{inspect(fires)}"

    assert {:ok, true} = Honker.Scheduler.resume(db, "due")
    {:ok, fires2} = Honker.Scheduler.tick(db)
    assert length(fires2) >= 1, "resumed schedule should emit; got #{inspect(fires2)}"
  end

  test "get_job misses after ack (separate from cancel)", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    {:ok, id} = Honker.Queue.enqueue(db, "emails", %{"to" => "x"})
    {:ok, job} = Honker.Queue.claim_one(db, "emails", "worker-1")
    assert {:ok, true} = Honker.Job.ack(db, job)
    assert {:ok, nil} = Honker.Queue.get_job(db, id)
  end

  test "update payload null vs omitted distinction", ctx do
    {:ok, db} = Honker.open(ctx.db_path, extension_path: ctx.ext)
    :ok =
      Honker.Scheduler.add(db,
        name: "t",
        queue: "q",
        schedule: "0 9 * * *",
        payload: %{"v" => 1}
      )

    # Omitted payload — leaves alone.
    assert {:ok, true} = Honker.Scheduler.update(db, "t", priority: 7)
    {:ok, rows} = Honker.Scheduler.list(db)
    assert Jason.decode!(Enum.at(rows, 0)["payload"])["v"] == 1

    # payload: nil — explicitly write JSON null.
    assert {:ok, true} = Honker.Scheduler.update(db, "t", payload: nil)
    {:ok, rows} = Honker.Scheduler.list(db)
    assert Jason.decode!(Enum.at(rows, 0)["payload"]) == nil
  end
end

File: packages/honker-bun/test/phase_mantle.test.ts
// Phase Mantle: Scheduler lifecycle + Queue cancel/getJob.

import { test, expect, describe } from "bun:test";
import { existsSync, mkdtempSync, rmSync } from "node:fs";
import { tmpdir } from "node:os";
import { join, resolve } from "node:path";

import { open, Scheduler } from "../src/index.ts";

const REPO_ROOT = resolve(import.meta.dir, "..", "..", "..");
const EXT_CANDIDATES = [
  "target/release/libhonker_ext.dylib",
  "target/release/libhonker_ext.so",
  "target/release/libhonker_extension.dylib",
  "target/release/libhonker_extension.so",
];
function findExtension(): string | null {
  const fromEnv = process.env.HONKER_EXT_PATH;
  if (fromEnv && existsSync(fromEnv)) return fromEnv;
  for (const rel of EXT_CANDIDATES) {
    const p = join(REPO_ROOT, rel);
    if (existsSync(p)) return p;
  }
  return null;
}

const extPath = findExtension();
if (!extPath && process.env.CI) {
  throw new Error("HONKER_EXT_PATH not found in CI; Bun tests must run for real");
}
const maybe = extPath ? describe : describe.skip;

maybe("phase mantle", () => {
  function withDb(fn: (db: ReturnType<typeof open>) => void) {
    const dir = mkdtempSync(join(tmpdir(), "honker-bun-mantle-"));
    const dbPath = join(dir, "t.db");
    const db = open(dbPath, extPath!);
    try {
      fn(db);
    } finally {
      db.close();
      rmSync(dir, { recursive: true, force: true });
    }
  }

  test("schedule list round-trips fields", () => {
    withDb((db) => {
      const sched = new Scheduler(db);
      sched.add({ name: "recap", queue: "emails", cron: "0 9 * * 1", payload: { team: "premier-league" }, priority: 3 });
      sched.add({ name: "sync", queue: "syncs", cron: "@every 1h", payload: null });

      const rows = sched.list();
      expect(rows.length).toBe(2);
      const recap = rows.find((r) => r.name === "recap")!;
      expect(recap.queue).toBe("emails");
      expect(recap.priority).toBe(3);
      expect(recap.enabled).toBe(true);
      expect(JSON.parse(recap.payload).team).toBe("premier-league");
    });
  });

  test("pause / resume idempotent", () => {
    withDb((db) => {
      const sched = new Scheduler(db);
      sched.add({ name: "a", queue: "q", cron: "0 9 * * *", payload: null });

      expect(sched.pause("a")).toBe(true);
      expect(sched.pause("a")).toBe(false);
      expect(sched.pause("missing")).toBe(false);
      expect(sched.list().find((r) => r.name === "a")!.enabled).toBe(false);

      expect(sched.resume("a")).toBe(true);
      expect(sched.resume("a")).toBe(false);
      expect(sched.list().find((r) => r.name === "a")!.enabled).toBe(true);
    });
  });

  test("update mutates fields, no-op on empty, recomputes next_fire_at on cron", () => {
    withDb((db) => {
      const sched = new Scheduler(db);
      sched.add({ name: "t", queue: "q", cron: "0 9 * * *", payload: { v: 1 } });

      expect(sched.update("t", { payload: { v: 99 }, priority: 5 })).toBe(true);
      const row = sched.list().find((r) => r.name === "t")!;
      expect(JSON.parse(row.payload).v).toBe(99);
      expect(row.priority).toBe(5);

      const before = row.next_fire_at;
      expect(sched.update("t", { cron: "*/5 * * * *" })).toBe(true);
      const after = sched.list().find((r) => r.name === "t")!;
      expect(after.cron_expr).toBe("*/5 * * * *");
      expect(after.next_fire_at).not.toBe(before);

      expect(sched.update("t")).toBe(false); // no-op
      expect(sched.update("missing", { payload: {} })).toBe(false);
    });
  });

  test("queue cancel + getJob", () => {
    withDb((db) => {
      const q = db.queue("emails");
      const id = q.enqueue({ to: "alice" });
      const row = q.getJob(id);
      expect(row).not.toBeNull();
      expect(row!.id).toBe(id);
      expect(row!.state).toBe("pending");

      expect(q.cancel(id)).toBe(true);
      expect(q.cancel(id)).toBe(false);
      expect(q.getJob(id)).toBeNull();
      expect(q.claimOne("worker-1")).toBeNull();
    });
  });

  test("cancel of processing invalidates ack", () => {
    withDb((db) => {
      const q = db.queue("emails");
      const id = q.enqueue({ to: "x" });
      const job = q.claimOne("worker-1")!;
      expect(job.id).toBe(id);
      expect(q.cancel(id)).toBe(true);
      expect(job.ack()).toBe(false);
    });
  });

  test("paused schedule does not emit on tick", async () => {
    let _db: any;
    const dir = mkdtempSync(join(tmpdir(), "honker-bun-mantle-"));
    const dbPath = join(dir, "t.db");
    const db = open(dbPath, extPath!);
    try {
      const sched = new Scheduler(db);
      sched.add({ name: "due", queue: "emails", cron: "@every 1s", payload: { x: 1 } });
      await new Promise((r) => setTimeout(r, 1100));
      sched.pause("due");
      const future = Math.floor(Date.now() / 1000) + 5;
      const fires = sched.tick(future);
      expect(fires.length).toBe(0);
      sched.resume("due");
      const fires2 = sched.tick(future);
      expect(fires2.length).toBeGreaterThanOrEqual(1);
    } finally {
      db.close();
      rmSync(dir, { recursive: true, force: true });
    }
  });

  test("get_job misses after ack (separate from cancel)", () => {
    withDb((db) => {
      const q = db.queue("emails");
      const id = q.enqueue({ to: "x" });
      const job = q.claimOne("worker-1")!;
      expect(job.ack()).toBe(true);
      expect(q.getJob(id)).toBeNull();
    });
  });

  test("update payload null vs omitted distinction", () => {
    withDb((db) => {
      const sched = new Scheduler(db);
      sched.add({ name: "t", queue: "q", cron: "0 9 * * *", payload: { v: 1 } });
      // Omitted payload — leaves alone.
      sched.update("t", { priority: 7 });
      let row = sched.list().find((r) => r.name === "t")!;
      expect(JSON.parse(row.payload).v).toBe(1);
      expect(row.priority).toBe(7);
      // payload: null — write JSON null.
      sched.update("t", { payload: null });
      row = sched.list().find((r) => r.name === "t")!;
      expect(JSON.parse(row.payload)).toBeNull();
    });
  });
});
